docs/proposals/pepxxx_wheel_variant_support.md

```diff
@@ -973,35 +973,42 @@ needs to be forked into a new package.
 To make it easier to discover and install plugins, they should be published in the same indexes that the packages using
 them. In particular, packages published to PyPI must not rely on plugins that need to be installed from other indexes
 
-Plugins are implemented using a Python class. Their API can be accessed using two methods:
+Plugins are implemented as Python modules. The API specified in this PEP can either be implemented as top-level
+attributes and functions, or as members of a class. In the latter case, the class is instantiated prior to accessing
+them.
 
-1. An explicit object reference, as the "plugin API" metadata.
+#### API endpoint
 
-2. An installed entry point in the variant_plugins group. The name of the entry point is insignificant, and the value
-provides the object reference.
+The location of the plugin code is called an "API endpoint", and it is expressed using the object reference notation
+following the [entry point specification](https://packaging.python.org/en/latest/specifications/entry-points/).
+They are in the form of:
 
-Both formats use the object reference notation from the
-[entry point specification](https://packaging.python.org/en/latest/specifications/entry-points/). That is, they are in
-the form of:
-
-```python
-importable.module:ClassName
+```
+{import path}(:{object path})?
 ```
 
-The resulting plugin is instantiated by the equivalent of:
+An API endpoint specification is equivalent to the following Python pseudocode:
 
 ```python
-import importable.module
+import {import path}
 
-plugin_instance = importable.module.ClassName()
+if "{object path}":
+    plugin = {import path}.{object path}
+else:
+    plugin = {import path}
+
+if callable(plugin):
```

Comment:
This will fail if callable modules should be introduced (https://peps.python.org/pep-0713/), we should use a real type check here.

Reply:
Do you have a specific suggestion?

Reply:
We can use the inspect module:

```python
import inspect
import os


class Foo:
    pass


print(
    inspect.ismodule(os),
    inspect.ismodule(Foo),
    inspect.isclass(os),
    inspect.isclass(Foo),
)
```

Reply:
Production implementations will want to use `isinstance(object, types.ModuleType)`/`isinstance(object, type)` directly to avoid the import overhead.

Reply:
Can you think of other types that could technically meet "plugin API" prototype besides a module (that should not be called) or a class (that needs to be instantiated)?

Perhaps it's time to remove class instantiation support altogether. We didn't really come up with any use for stateful plugins, and with plugins being called in isolation the concept becomes even more useless.

Reply:
Just a module seems simpler.

PEP 517 doesn't care at all what type it gets, the following works:

```toml
[build-system]
requires = ["hatchling"]
backend-path = ["."]
build-backend = "backend:foo"
```

```python
import hatchling.build


def foo():
    pass


foo.build_wheel = hatchling.build.build_wheel
foo.build_sdist = hatchling.build.build_sdist
```

Reply:
Oh my, that snippet is a true nightmare. I suppose `inspect.isclass()` is good enough for it now, though.